Warn user once for possible unnecessary find_unused_params in DDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50133 Warn user once for possible unnecessary find_unused_params in DDP**

`find_unused_parameters=True` is only needed when the model has unused parameters that are not known at model definition time or differ due to control flow.

Unfortunately, many DDP users pass this flag in as `True` even when they do not need it, sometimes as a precaution to mitigate possible errors that may be raised (such as the error we raise with not using all outputs).While this is a larger issue to be fixed in DDP, it would also be useful to warn once if we did not detect unused parameters.

The downside of this is that in the case of flow control models where the first iteration doesn't have unused params but the rest do, this would be a false warning. However, I think the warning's value exceeds this downside, and we have also indicated this possibility in the warning itself.

Differential Revision: [D25411118](https://our.internmc.facebook.com/intern/diff/D25411118/)

rohan-varma committed:
warn user once for possible unnecessary find_unused_params

`find_unused_parameters=True` is only needed when the model has unused parameters that are not known at model definition time or differ due to control flow.

Unfortunately, many DDP users pass this flag in as `True` even when they do not need it, sometimes as a precaution to mitigate possible errors that may be raised (such as the error we raise with not using all outputs).While this is a larger issue to be fixed in DDP, it would also be useful to warn once if we did not detect unused parameters.

The downside of this is that in the case of flow control models where the first iteration doesn't have unused params but the rest do, this would be a false warning. However, I think the warning's value exceeds this downside.

Differential Revision: [D25411118](https://our.internmc.facebook.com/intern/diff/D25411118/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1036,6 +1036,22 @@ void Reducer::prepare_for_backward(
   }
+
+  // Warn user about unnecessary perf hit if all parameters were used.
+  if (unused_parameters_.empty() && !foundUnusedParametersOnce_) {
```
Do we need the additional variable `foundUnusedParametersOnce_`? Shouldn't `TORCH_WARN_ONCE` take care of this?

rohan-varma replied:
Yes it should, thanks for the catch

rohan-varma commented:
Modified a UT that uses all parameters and verified we see the logging only once when there are many iterations:
```
WARNING: Logging before InitGoogleLogging() is written to STDERR
I0107 16:04:22.672833 3885185 ProcessGroupNCCL.cpp:542] [Rank 1] NCCL watchdog thread started!
I0107 16:04:22.672832 3884645 ProcessGroupNCCL.cpp:473] [Rank 1] ProcessGroupNCCL initialized with following options:
NCCL_ASYNC_ERROR_HANDLING: 0
NCCL_BLOCKING_WAIT: 0
TIMEOUT(ms): 60000
USE_HIGH_PRIORITY_STREAM: 0
WARNING: Logging before InitGoogleLogging() is written to STDERR
I0107 16:04:22.679401 3885187 ProcessGroupNCCL.cpp:542] [Rank 0] NCCL watchdog thread started!
I0107 16:04:22.679401 3884644 ProcessGroupNCCL.cpp:473] [Rank 0] ProcessGroupNCCL initialized with following options:
NCCL_ASYNC_ERROR_HANDLING: 0
NCCL_BLOCKING_WAIT: 0
TIMEOUT(ms): 60000
USE_HIGH_PRIORITY_STREAM: 0
NCCL version 2.7.3+cudaCUDA_MAJOR.CUDA_MINOR
W0107 16:04:26.764827 3884645 reducer.cpp:1050] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters. This flag results in an extra traversal of the autograd graph every iteration, which can adversely affect performance. If your model indeed never has any unused parameters, consider turning this flag off. Note that this warning may be a false positive your model has flow control causing later iterations to have unused parameters. (function operator())
W0107 16:04:26.768517 3884644 reducer.cpp:1050] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters. This flag results in an extra traversal of the autograd graph every iteration, which can adversely affect performance. If your model indeed never has any unused parameters, consider turning this flag off. Note that this warning may be a false positive your model has flow control causing later iterations to have unused parameters. (function operator())
I0107 16:04:26.795979 3885187 ProcessGroupNCCL.cpp:544] [Rank 0] NCCL watchdog thread terminated normally
I0107 16:04:26.892848 3885185 ProcessGroupNCCL.cpp:544] [Rank 1] NCCL watchdog thread terminated normally
ok
```

rohan-varma committed:
Update on "Warn user once for possible unnecessary find_unused_params in DDP"


`find_unused_parameters=True` is only needed when the model has unused parameters that are not known at model definition time or differ due to control flow.

Unfortunately, many DDP users pass this flag in as `True` even when they do not need it, sometimes as a precaution to mitigate possible errors that may be raised (such as the error we raise with not using all outputs).While this is a larger issue to be fixed in DDP, it would also be useful to warn once if we did not detect unused parameters.

The downside of this is that in the case of flow control models where the first iteration doesn't have unused params but the rest do, this would be a false warning. However, I think the warning's value exceeds this downside, and we have also indicated this possibility in the warning itself.

Differential Revision: [D25411118](https://our.internmc.facebook.com/intern/diff/D25411118/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Warn user once for possible unnecessary find_unused_params in DDP"


`find_unused_parameters=True` is only needed when the model has unused parameters that are not known at model definition time or differ due to control flow.

Unfortunately, many DDP users pass this flag in as `True` even when they do not need it, sometimes as a precaution to mitigate possible errors that may be raised (such as the error we raise with not using all outputs).While this is a larger issue to be fixed in DDP, it would also be useful to warn once if we did not detect unused parameters.

The downside of this is that in the case of flow control models where the first iteration doesn't have unused params but the rest do, this would be a false warning. However, I think the warning's value exceeds this downside, and we have also indicated this possibility in the warning itself.

Differential Revision: [D25411118](https://our.internmc.facebook.com/intern/diff/D25411118/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Warn user once for possible unnecessary find_unused_params in DDP"


`find_unused_parameters=True` is only needed when the model has unused parameters that are not known at model definition time or differ due to control flow.

Unfortunately, many DDP users pass this flag in as `True` even when they do not need it, sometimes as a precaution to mitigate possible errors that may be raised (such as the error we raise with not using all outputs).While this is a larger issue to be fixed in DDP, it would also be useful to warn once if we did not detect unused parameters.

The downside of this is that in the case of flow control models where the first iteration doesn't have unused params but the rest do, this would be a false warning. However, I think the warning's value exceeds this downside, and we have also indicated this possibility in the warning itself.

Differential Revision: [D25411118](https://our.internmc.facebook.com/intern/diff/D25411118/)

[ghstack-poisoned]
